sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/EnvironmentConfiguration.java

```diff
@@ -0,0 +1,241 @@
+ * used during construction of client libraries.
+ */
+public class EnvironmentConfiguration implements Cloneable {
+    /*
+     * Configurations that are loaded into the global configuration store when the application starts.
+     */
+    private static final Set<String> DEFAULT_CONFIGURATIONS = new HashSet<>(Arrays.asList(
+        Configuration.PROPERTY_HTTP_PROXY,
+        Configuration.PROPERTY_HTTPS_PROXY,
+        Configuration.PROPERTY_IDENTITY_ENDPOINT,
+        Configuration.PROPERTY_IDENTITY_HEADER,
+        Configuration.PROPERTY_NO_PROXY,
+        Configuration.PROPERTY_MSI_ENDPOINT,
+        Configuration.PROPERTY_MSI_SECRET,
+        Configuration.PROPERTY_AZURE_SUBSCRIPTION_ID,
+        Configuration.PROPERTY_AZURE_USERNAME,
+        Configuration.PROPERTY_AZURE_PASSWORD,
+        Configuration.PROPERTY_AZURE_CLIENT_ID,
+        Configuration.PROPERTY_AZURE_CLIENT_SECRET,
+        Configuration.PROPERTY_AZURE_TENANT_ID,
+        Configuration.PROPERTY_AZURE_CLIENT_CERTIFICATE_PATH,
+        Configuration.PROPERTY_AZURE_IDENTITY_DISABLE_CP1,
+        Configuration.PROPERTY_AZURE_RESOURCE_GROUP,
+        Configuration.PROPERTY_AZURE_CLOUD,
+        Configuration.PROPERTY_AZURE_AUTHORITY_HOST,
+        Configuration.PROPERTY_AZURE_TELEMETRY_DISABLED,
+        Configuration.PROPERTY_AZURE_LOG_LEVEL,
+        Configuration.PROPERTY_AZURE_HTTP_LOG_DETAIL_LEVEL,
+        Configuration.PROPERTY_AZURE_TRACING_DISABLED,
+        Configuration.PROPERTY_AZURE_POD_IDENTITY_TOKEN_URL,
+        Configuration.PROPERTY_AZURE_REGIONAL_AUTHORITY_NAME,
+        Configuration.PROPERTY_AZURE_REQUEST_RETRY_COUNT,
+        Configuration.PROPERTY_AZURE_REQUEST_CONNECT_TIMEOUT,
+        Configuration.PROPERTY_AZURE_REQUEST_WRITE_TIMEOUT,
+        Configuration.PROPERTY_AZURE_REQUEST_RESPONSE_TIMEOUT,
+        Configuration.PROPERTY_AZURE_REQUEST_READ_TIMEOUT
+    ));
+
+    private static final EnvironmentConfiguration GLOBAL_CONFIGURATION = new EnvironmentConfiguration();
+
+    private final ConcurrentMap<String, Optional<String>> configurations;
+
+    /**
+     * Constructs a configuration containing the known Azure properties constants.
+     */
+    private EnvironmentConfiguration() {
+        this.configurations = loadBaseConfiguration();
+    }
+
+    /**
+     * Constructs a configuration containing mocked environment. Use this constructor for testing.
+     */
+    public EnvironmentConfiguration(Map<String, String> configurations) {
+        Objects.requireNonNull(configurations, "'configurations' can't be null");
+        this.configurations = new ConcurrentHashMap<>(configurations.size());
+        for (Map.Entry<String, String> config : configurations.entrySet()) {
+            this.configurations.put(config.getKey(), Optional.ofNullable(config.getValue()));
+        }
+    }
+
+    private EnvironmentConfiguration(EnvironmentConfiguration original) {
```

Comment:
Oh I see the copy constructor, so, we don't really need the Cloneable interface.